rfcs/000X-flux-bootstrap-oci/README.md

```diff
@@ -0,0 +1,319 @@
+kind: OCIRepository
+metadata:
+  name: flux-components
+  namespace: flux-system
+spec:
+  interval: 1m
+  url: oci://ghcr.io/stefanprodan/flux-manifests
+  ref:
+    tag: production
+  secretRef:
+    name: flux-components
+---
+apiVersion: kustomize.toolkit.fluxcd.io/v1
+kind: Kustomization
+metadata:
+  name: flux-components
+  namespace: flux-system
+spec:
+    interval: 1h
+    retryInterval: 5m
+    sourceRef:
+        kind: OCIRepository
+        name: flux-components
+    path: ./
+    prune: true
+```
+
+Cluster state reconciliation:
+
+```yaml
+apiVersion: source.toolkit.fluxcd.io/v1beta2
+kind: OCIRepository
+metadata:
+  name: flux-system
+  namespace: flux-system
+spec:
+   interval: 1m
+   url: oci://ghcr.io/stefanprodan/fleet-manifests
+   ref:
+    tag: production
+   secretRef:
+    name: flux-system
+---
+apiVersion: kustomize.toolkit.fluxcd.io/v1
+kind: Kustomization
+metadata:
+  name: flux-system
+  namespace: flux-system
+spec:
+    interval: 1h
+    retryInterval: 5m
+    sourceRef:
+        kind: OCIRepository
+        name: flux-system
+    path: clusters/production
+    prune: true
+```
+
+#### Story 2
+
+> As a platform operator I want to sync the cluster state with the fleet Git repository.
+
+Push changes from the fleet Git repository to the container registry:
+
+```shell
+# clone the fleet Git repository
+git clone https://github.com/stefanprodan/fleet.git
+cd fleet
+git switch main
+
+# push the contents the fleet OCI repository and tag it with the commit short SHA
+flux push artifact oci://ghcr.io/stefanprodan/fleet-manifests:$(git rev-parse --short HEAD) \
+--path="./" \
+--source="$(git config --get remote.origin.url)" \
+--revision="$(git branch --show-current)@sha1:$(git rev-parse HEAD)"
+
+# tag the new version for production
+flux tag artifact oci://ghcr.io/stefanprodan/fleet-manifests:$(git rev-parse --short HEAD) \
+--tag=production
+```
+
+This operation can be automated using the Flux GitHub Action.
+
+The Git repository structure would be similar to the
+[flux2-kustomize-helm-example](https://github.com/fluxcd/flux2-kustomize-helm-example) with the following changes:
+
+- The `clusters/production/flux-system` directory is no more.
+- The Flux Kustomization objects defined in the `clusters/production` directory, such as
+  `infrastructure.yaml` and `apps.yaml`, have the `.spec.sourceRef` set to
+  `kind: OCIRepository` and  `name: flux-system`.
+
+#### Story 3
+
+> As a platform operator I want to update the Flux controllers on my production cluster
+> from CI without access to the Kubernetes API.
+
+Download the latest CLI version and update Flux directly in the registry, without rerunning bootstrap:
+
+```shell
+# pull the latest manifests from the registry
+flux pull artifact oci://ghcr.io/stefanprodan/flux-manifests:production \
+--output=./flux-manifests
+
+# update the Flux components manifests
+flux install --export > ./flux-manifests/flux-system/gotk-components.yaml
```

Comment:
These are two alternative methods right? It's not very clear from the text at the moment.

Reply:
If you don't have access to the cluster (what the user story is about), this is the only way. If you have API access, then like with Git bootstrap, you can just rerun it to update. OCI bootstrap behaves the same as Git bootstrap.

Reply:
I just thought these two commands will write the same kind of output, except that install lets you select a subset of controllers... maybe I am missing something else.

Reply:
Is it the case right now that one has to rerun bootstrap on major/minor releases while patch releases are taken care of by in-cluster image version bumps?

Reply:
Every time we release Flux, users get an PR opened to update the manifests in Git. For OCI you would need some kind of semver range or some other manual gate e.g. a GitHub workflow dispatch to approve minor bumps and let only patch versions be automatically push to the registry.

Reply:
> I just thought these two commands will write the same kind of output, except that install lets you select a subset of controllers... maybe I am missing something else.

All the `flux bootstrap` commands have the same args as `flux install`, so you can pick controllers, etc with bootstrap too. If you bootstrap with `flux bootstrap oci --components --log-level=source-controller,kustomize-controller`, to update your would run `flux install --log-level=source-controller,kustomize-controller --export` in CI.

Reply:
I mean 'pull artifact' vs 'install --export'.

Reply:
You need to pull to preserve the existing kustomization.yaml and any other extra resources you may have added at bootstrap. The `install -export` command only generates the components YAML.

Reply:
Ah, of course, this is a rebase ;)
Maybe the text needs to make that clearer.